### PR TITLE
Finish up type hints for federation client code

### PR DESCRIPTION
**Pull request by clokep**
Requires type hints for `synapse/federation/transport/client.py` and `synapse/http/matrixfederationclient.py`. The commits are fairly separate.

### CODE REVIEW

File: synapse/federation/federation_client.py [hunk 1]
```diff
@@ -280,15 +280,11 @@ async def backfill(
         logger.debug("backfill transaction_data=%r", transaction_data)
 
         if not isinstance(transaction_data, dict):
-            # TODO we probably want an exception type specific to federation
-            # client validation.
-            raise TypeError("Backfill transaction_data is not a dict.")
+            raise InvalidResponseError("Backfill transaction_data is not a dict.")
```
Comment (clokep) [inline synapse/federation/federation_client.py:283]: This is mostly a drive-by, but we seem to have added a specific type for this at some point...

File: synapse/federation/transport/client.py [hunk 2]
```diff
@@ -388,6 +404,7 @@ async def send_leave_v1(
             # server was just having a momentary blip, the room will be out of
             # sync.
             ignore_backoff=True,
+            parser=LegacyJsonDictParser(),
```
Comment (clokep) [inline synapse/federation/transport/client.py:407]: The v1 versions of `send_{leave,join}` have a weird return type of `[status code, actual response]`.

File: synapse/http/matrixfederationclient.py [hunk 3]
```diff
@@ -902,7 +935,7 @@ async def put_json(
             _sec_timeout = self.default_timeout
 
         if parser is None:
-            parser = JsonParser()
+            parser = cast(ByteParser[T], JsonParser())
```
Comment (clokep) [inline synapse/http/matrixfederationclient.py:938]: mypy didn't like not having this cast:

```
synapse/http/matrixfederationclient.py:938: error: Incompatible types in assignment (expression has type "JsonParser", variable has type "Optional[ByteParser[T]]")  [assignment]
synapse/http/matrixfederationclient.py:946: error: Argument "parser" to "_handle_response" has incompatible type "Optional[ByteParser[T]]"; expected "ByteParser[T]"  [arg-type]
```

File: synapse/http/matrixfederationclient.py [hunk 4]
```diff
@@ -924,7 +957,7 @@ async def post_json(
         timeout: Optional[int] = None,
         ignore_backoff: bool = False,
         args: Optional[QueryParams] = None,
-    ) -> Union[JsonDict, list]:
+    ) -> JsonDict:
```
Comment (clokep) [inline synapse/http/matrixfederationclient.py:960]: All of the callers want dicts anyway, so I figured we should just do it here instead of in each one.

File: synapse/http/matrixfederationclient.py (hunk 3 above)
Comment (DMRobertson) [reply synapse/http/matrixfederationclient.py:940]: I would guess this is because JsonParser is a `ByteParser[Union[JsonDict, list]]`and not a `ByteParser[T]`. Mypy might be happier if we make two separate calls to `handle_response`. But I don't think we need to bother.

File: synapse/http/matrixfederationclient.py [hunk 5]
```diff
@@ -184,20 +184,54 @@ def get_json(self) -> Optional[JsonDict]:
...
+    def __init__(self) -> None:
+        super().__init__(self._validate)
+
+    @staticmethod
+    def _validate(v: Any) -> bool:
+        # Match [integer, JSON dict]
+        return (
+            isinstance(v, list)
+            and len(v) == 2
+            and type(v[0]) == int
```
Comment (DMRobertson) [inline synapse/http/matrixfederationclient.py:234]: I guess we check `type` rather than `isinstance` to explicitly reject floats?

File: synapse/http/matrixfederationclient.py [hunk 6]
```diff
@@ -184,20 +184,54 @@ def get_json(self) -> Optional[JsonDict]:
...
 
 
-class JsonParser(ByteParser[Union[JsonDict, list]]):
+class _BaseJsonParser(ByteParser[T]):
     """A parser that buffers the response and tries to parse it as JSON."""
 
     CONTENT_TYPE = "application/json"
 
-    def __init__(self) -> None:
+    def __init__(self, validator: Optional[Callable[[Any], bool]] = None) -> None:
```
Comment (DMRobertson) [inline synapse/http/matrixfederationclient.py:192]: Could we have the validator accept `object` rather than `Any`?  It won't make a difference here, but in general I prefer `object`, since `Any` effectively opts out of typechecking.

File: synapse/http/matrixfederationclient.py [hunk 7]
```diff
@@ -184,20 +184,54 @@ def get_json(self) -> Optional[JsonDict]:
...
 
     def write(self, data: bytes) -> int:
         return self._binary_wrapper.write(data)
 
-    def finish(self) -> Union[JsonDict, list]:
-        return json_decoder.decode(self._buffer.getvalue())
+    def finish(self) -> T:
+        result = json_decoder.decode(self._buffer.getvalue())
+        if self._validator is not None and not self._validator(result):
+            raise ValueError("Unexpected JSON object")
```
Comment (DMRobertson) [inline synapse/http/matrixfederationclient.py:203]: If JsonDictParser is given a list, it'll complain that it recieved an unexpected JSON object, which seems wrong. Should we include the `result` in the error message too?

Perhaps `f"Malformed JSON value: {result}"`?

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Thanks! My only objection is the error message in `finish`, the rest are just nice-to-haves.

File: synapse/http/matrixfederationclient.py (hunk 5 above)
Comment (clokep) [reply synapse/http/matrixfederationclient.py:234]: Explicitly reject `bool`

```pycon
>>> isinstance(1, int)
True
>>> isinstance(True, int)
True
>>> isinstance(1.0, int)
False
```

File: synapse/http/matrixfederationclient.py (hunk 6 above)
Comment (clokep) [reply synapse/http/matrixfederationclient.py:192]: I think so? It would need to be `Optional[object]` though, I think.

File: synapse/http/matrixfederationclient.py (hunk 7 above)
Comment (clokep) [reply synapse/http/matrixfederationclient.py:203]: > If JsonDictParser is given a list, it'll complain that it recieved an unexpected JSON object, which seems wrong. Should we include the `result` in the error message too?

Why is this wrong? A JSON array isn't an object.

Comment (clokep) [reply synapse/http/matrixfederationclient.py:203]: Do you mean that I abstracted the validator, but not the error message?

Comment (DMRobertson) [reply synapse/http/matrixfederationclient.py:203]: > > If JsonDictParser is given a list, it'll complain that it recieved an unexpected JSON object, which seems wrong. Should we include the `result` in the error message too?
> 
> Why is this wrong? A JSON array isn't an object.

If I expect a Dict/object and I get a List/array, the error message should say "unexpected array", not "unexpected object". Or better: "Expected object, but got an array". The current behaviour:

```python
>>> from synapse.http.matrixfederationclient import JsonDictParser
>>> p = JsonDictParser()
>>> p.write(b"[]")
2
>>> p.finish()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dmr/workspace/synapse/synapse/http/matrixfederationclient.py", line 203, in finish
    raise ValueError("Unexpected JSON object")
ValueError: Unexpected JSON object
```

> Do you mean that I abstracted the validator, but not the error message?

Not worried about abstracting out the error message, I just want to ensure the error message won't confuse our future selves.

Comment (clokep) [reply synapse/http/matrixfederationclient.py:203]:
```suggestion
            raise ValueError(f"Received incorrect JSON value: {result.__class__.__name__}")
```

Comment (DMRobertson) [reply synapse/http/matrixfederationclient.py]: Works for me (sorry for digging in on this). Looks like a test might be suddenly failing though?

**Comment by clokep** [comment]
@DMRobertson I made the change we discussed -- by default we now parse a JSON object and only use a separate parser if we have a different expectation (i.e. the `/send_{join,leave}` endpoints).

**Review by DMRobertson — APPROVED** [review]
Thanks for bearing with me.
